### PR TITLE
Eval loss

### PR DESCRIPTION
**Pull request by Sehmimul**
- Major refractoring in `diffusion.py` script
- Eval loss can now be obtained from the test_loader object

### CODE REVIEW

**Comment by vishnu-m77** [comment]
Can you pull the latest changes from test branch? It has confusion matrix, and changes to report + plots.

File: main.py [hunk 1]
```diff
@@ -74,7 +74,7 @@
         # Initialize DCG
         dcg = dcg_module.DCG(dcg_params)
         # Trains DCG and saves the model
-        dcg_module.train_DCG(dcg, dcg_params, train_loader)
+        dcg_module.train_DCG(dcg, dcg_params, train_loader, test_loader=test_loader)
```
Comment (vishnu-m77) [inline main.py:77]: Why are you passing test_loader into DCG?

File: src/DCG/main.py [hunk 2]
```diff
@@ -110,28 +110,41 @@ def cast_label_to_one_hot_and_prototype(self, y_labels_batch, return_prototype=T
...
-        loss_batch = []
+        # loss_batch = []
         for feature_label_set in train_loader:
+            # train loss
+            dcg.train()
             x_batch, y_labels_batch = feature_label_set
             y_one_hot_batch, y_logits_batch = dcg.cast_label_to_one_hot_and_prototype(y_labels_batch)
             aux_loss = dcg.nonlinear_guidance_model_train_step(criterion, x_batch, y_one_hot_batch, optimizer)
             loss_batch.append(aux_loss)
+            # eval loss
```
Comment (vishnu-m77) [inline src/DCG/main.py:131]: My assumption was that you will create a separate eval function for DCG

File: main.py (hunk 1 above)
Comment (Sehmimul) [reply main.py:77]: To compute eval loss

File: src/DCG/main.py (hunk 2 above)
Comment (Sehmimul) [reply src/DCG/main.py:131]: I don’t think there is any need to create separate eval function. I am just getting eval loss at each train iteration.

Comment (vishnu-m77) [reply src/DCG/main.py:131]: Ohh it's supposed to be validation. Do you think it will be better to create a val_loader instead of using test_loader?

**Review by vishnu-m77 — APPROVED** [review]
Fantastic! Looks good to merge to main.
